src/PrestaShopBundle/Twig/Extension/GridExtension.php

```diff
@@ -0,0 +1,199 @@
+<?php
+/**
+ * 2007-2018 PrestaShop.
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShopBundle\Twig\Extension;
+
+use RuntimeException;
+use Symfony\Component\Cache\Adapter\AdapterInterface;
+use Twig\Environment;
+use Twig\Extension\AbstractExtension;
+use Twig_SimpleFunction as SimpleFunction;
+
+/**
+ * Class GridExtension is responsible for providing grid helpers functions.
+ *
+ * - column_content(column, record, grid): renders column content based on column type.
+ * - column_header(column, grid): renders column header based on column type.
+ */
+class GridExtension extends AbstractExtension
+{
+    const BASE_COLUMN_CONTENT_TEMPLATE_PATH = '@PrestaShop/Admin/Common/Grid/Columns/Content';
+    const BASE_COLUMN_HEADER_TEMPLATE_PATH = '@PrestaShop/Admin/Common/Grid/Columns/Header/Content';
+
+    /**
+     * @var Environment
+     */
+    private $twig;
+
+    /**
+     * @var AdapterInterface
+     */
+    private $cache;
+
+    /**
+     * @param Environment $twig
+     * @param AdapterInterface $cache
+     */
+    public function __construct(Environment $twig, AdapterInterface $cache)
+    {
+        $this->twig = $twig;
+        $this->cache = $cache;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getFunctions()
+    {
+        return [
+            new SimpleFunction('column_content', [$this, 'renderColumnContent'], [
+                'is_safe' => ['html'],
+            ]),
+            new SimpleFunction('column_header', [$this, 'renderColumnHeader'], [
+                'is_safe' => ['html'],
+            ]),
+        ];
+    }
+
+    /**
+     * Render column content.
+     *
+     * @param array $record
+     * @param array $column
+     * @param array $grid
+     *
+     * @return string
+     *
+     * @throws RuntimeException when template cannot be found for column
+     */
+    public function renderColumnContent(array $record, array $column, array $grid)
+    {
+        $templateCacheKey = sprintf('column_%s_%s_%s_content', $grid['id'], $column['id'], $column['type']);
```

Comment:
What about adding defensive validation here ?
```
$this->validateInput($record, $column, $grid);

private function validateInput(array $record, array $column, array $grid)
{
    $this->arrayMustContains($grid, 'id');
    $this->arrayMustContains($column, ['id', 'type]);
}

// might be an util function so could come from another class
/**
 * @param array $array
 * @param string|string[] $keys
 *
 * @throws \InvalidArgumentException
 */
protected function arrayMustContains(array $array, $keys)
{
    ...
}
```

Reply:
I don't want the validation done directly in the class (this breaks SRP).

Note that we have complete control on grid, column, and record (see GridPresenter class).

Reply:
We do that later :)